### ISSUE TITLE
How to solve the YOLOv8-cls class imbalance problem

### ISSUE THREAD

**Issue by kimdodo97** [ultralytics/ultralytics#2208]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am using the YOLOv8 classification model.
In the model, the classification is classifying a total of 5 labels.
Here I have about 1800,500,1200,5000,20000 data points for each label.
I expect to encounter class imbalance issues in this classification model, and I was wondering if a method like class balanced loss is currently included in the YOLOv8-cls model?

**Comment by glenn-jocher** [comment]
@jungms4536 thank you for your question. Yes, we do have class-balanced loss available in our YOLOv8-cls model. Class imbalance can indeed be an issue in classification tasks, and using class-balanced loss can help address this problem. You may also refer to our documentation at https://docs.ultralytics.com/ for more information on how to use this feature. Let us know if you have any further questions or concerns.

**Comment by jungms4536** [comment]
I also checked that the focal loss function is defined in yolov8, but I don't see anything in the documentation on how to use class balance loss, if you have a reference or method, could you please let me know?

**Comment by glenn-jocher** [comment]
@motidil sure, I'd be glad to guide you through the process. The loss function including the class-balanced loss is controlled in the YOLOv8 model configuration file (model.yaml). Within this file, you can modify the loss parameters to implement a class-balanced loss. You will find the section for "loss" in the model configuration file where you can adjust as needed.

Please remember to be careful when adjusting the loss settings as they can have a fundamental impact on your model's performance. Always check your results thoroughly to ensure the modifications are improving your model's prediction accuracy.

Let me know if you need any further assistance with the configuration settings.

**Comment by motidil** [comment]
@glenn-jocher are you sure the `model.yaml` exists for classify task? and specifically for `yolov8-cls`...where can I find it?

**Comment by glenn-jocher** [comment]
@motidil, I apologize if there was any confusion in my previous response. You are right, the `model.yaml` is typically more relevant for YOLO models focused on object detection tasks. In case of classification tasks, the class imbalance issue is often handled during the data preprocessing stage or within the loss function during model training. Currently, YOLOv8 applies focal loss which inherently provides a form of class balancing.

I'm sorry for any confusion - to clarify, there's no direct method to apply class balanced loss in `YOLOv8-cls` model specifically, but the model works to address class imbalance by using focal loss, which places more emphasis on the hard, misclassified examples. 

I hope this clears up the confusion. Please let me know if you have further questions or concerns.

**Comment by MadMenHitBooker** [comment]
So what is the solution ? having equilibrated dataset from the beggining? Maybe using some library like albumentations ?

**Comment by motidil** [comment]
@MadMenHitBooker Yes, i used albumentations to enrich the data set

**Comment by XiFromEY** [comment]
hi @glenn-jocher, are there any sample code regarding the imbalance problem?

**Comment by glenn-jocher** [comment]
@XiFromEY class imbalance is a common issue in machine learning and can significantly affect the performance of classification models. While YOLOv8-cls does not have a specific "class balanced loss" function, there are several strategies you can employ to mitigate class imbalance:

1. **Weighted Loss Function**: Assign different weights to classes in the loss function based on the inverse frequency of the classes. This helps to give more importance to minority classes during training.

2. **Data Augmentation**: Augment the data of minority classes to increase their representation in the dataset. Techniques like flipping, rotation, scaling, and cropping can help generate more data for underrepresented classes.

3. **Oversampling**: Oversample the minority classes by duplicating existing samples or generating new samples through data augmentation.

4. **Undersampling**: Undersample the majority classes by reducing the number of samples to balance the dataset.

5. **Focal Loss**: Use a focal loss function, which is designed to down-weight the loss assigned to well-classified examples and focus more on hard, misclassified examples.

6. **Transfer Learning**: Use transfer learning to fine-tune a pre-trained model on your dataset. This can help the model generalize better, even with imbalanced data.

7. **Custom Loss Function**: Implement a custom loss function that takes class imbalance into account.

To implement a weighted loss function in YOLOv8-cls, you can modify the training script to include class weights in the loss calculation. Here's a conceptual example of how you might adjust the loss function:

```python
# Calculate class weights (inverse frequency or another method)
class_counts = [1800, 500, 1200, 5000, 20000]
total_counts = sum(class_counts)
class_weights = [total_counts / count for count in class_counts]

# Incorporate class weights into the loss function
# This is a conceptual example and may require adaptation to the actual YOLOv8 codebase
def weighted_cross_entropy_loss(predictions, targets):
    loss = 0
    for i, (pred, target) in enumerate(zip(predictions, targets)):
        weight = class_weights[target]
        loss += weight * cross_entropy(pred, target)
    return loss
```

In practice, you would need to integrate this into the YOLOv8-cls training pipeline, ensuring that the loss function is used during the backward pass and that the weights are properly normalized.

If you're not familiar with modifying the loss function or if it seems too complex, starting with data-level techniques like oversampling or augmentation might be a more accessible approach.

Remember to validate your model thoroughly after applying these techniques to ensure that the performance has improved and that the model is not overfitting to the minority classes.

**Comment by XiFromEY** [comment]
@glenn-jocher  thank you for your reply, what is training script? How can I incorporate into the current code? I am using xView dataset which is fixed so I cannot do any data-level techniques (the damaged buildings' number is so small):(

**Comment by glenn-jocher** [comment]
@XiFromEY class imbalance is a common issue in machine learning and can significantly affect the performance of classification models. While YOLOv8-cls does not have a specific "class balanced loss" function built-in, there are several strategies you can employ to mitigate class imbalance issues:

1. **Weighted Loss Function**: Assign different weights to the classes in the loss function based on the inverse frequency of the classes. This can help the model pay more attention to the minority classes.

2. **Data Augmentation**: Augment the data for the minority classes to increase their representation in the dataset. This can include techniques like oversampling, image transformations, or generating synthetic data.

3. **Focal Loss**: Use a focal loss function, which is designed to focus more on hard-to-classify examples and can naturally help with class imbalance by down-weighting the loss assigned to well-classified examples.

4. **Transfer Learning**: Use a pretrained model and fine-tune it on your dataset. Pretrained models have learned features from a large and diverse dataset, which can help in learning representations for minority classes.

5. **Ensemble Methods**: Train multiple models on different subsets of the data and combine their predictions. This can include methods like bagging or boosting.

6. **Custom Loss Function**: Implement a custom loss function that takes class imbalance into account. You can modify the YOLOv8 code to include a loss function that better suits your needs.

7. **Post-processing**: Adjust the decision threshold for each class based on the validation set performance to improve precision and recall for the minority classes.

8. **Evaluation Metrics**: Use evaluation metrics that account for class imbalance, such as weighted F1-score, Matthews correlation coefficient, or area under the precision-recall curve (AUPRC), rather than metrics like accuracy.

To implement some of these strategies, you may need to modify the YOLOv8 codebase or preprocess your data accordingly. If you're interested in using a weighted loss function or focal loss, you would need to customize the loss computation in the model's training loop.

For example, to implement a simple weighted loss function, you could modify the classification loss part of the code to weigh the loss of each class by the inverse of its frequency. This would require you to calculate the weights based on your dataset's class distribution and then apply these weights within the loss function.

Remember that handling class imbalance often requires experimentation with different strategies to find the best approach for your specific dataset and use case.

**Comment by glenn-jocher** [comment]
@XiFromEY class imbalance is a common issue in machine learning and can significantly affect the performance of classification models. While YOLOv8-cls does not have a specific "class balanced loss" function built-in, there are several strategies you can employ to mitigate class imbalance:

1. **Weighted Loss Function**: Assign different weights to classes in the loss function based on the inverse frequency of the classes. This helps to give more importance to minority classes during training.

2. **Data Augmentation**: Augment the data of minority classes to increase their representation in the dataset. Techniques like flipping, rotation, scaling, and cropping can help generate more varied samples.

3. **Oversampling**: Oversample the minority classes by duplicating existing samples or generating new synthetic samples using techniques like SMOTE (Synthetic Minority Over-sampling Technique).

4. **Undersampling**: Undersample the majority classes by reducing the number of samples to balance the dataset.

5. **Focal Loss**: Use a focal loss function, which is designed to focus more on hard-to-classify examples and less on easy ones, effectively giving more weight to minority classes.

6. **Custom Loss Function**: Implement a custom loss function that takes class imbalance into account, such as the class balanced loss you mentioned.

7. **Two-Stage Training**: Train the model in two stages, where the first stage focuses on learning general features and the second stage focuses on the classification task with class balancing techniques applied.

8. **Transfer Learning**: Use transfer learning from a pre-trained model to help the model learn features that are not specific to the class distribution of your dataset.

9. **Ensemble Methods**: Combine the predictions of multiple models trained on different resampled datasets (e.g., bagging).

10. **Evaluation Metrics**: Use evaluation metrics that are less sensitive to class imbalance, such as F1-score, precision-recall curves, or the area under the ROC curve.

To implement some of these strategies in YOLOv8-cls, you may need to modify the training pipeline or loss functions. For example, to use weighted loss, you can modify the loss computation to include class weights. For data augmentation, you can use the built-in augmentation options or create a custom data loader that performs the desired augmentations.

If you're interested in implementing a specific technique like class balanced loss, you would need to customize the loss function in the YOLOv8 codebase. This would involve defining a new loss that accounts for the effective number of samples per class and integrating it into the training loop.

Remember to carefully validate your model after applying these techniques to ensure that the performance has improved and that the model is not overfitting to the minority classes.

**Comment by XiFromEY** [comment]
@glenn-jocher， can you please clarify how to implement a specific technique like class balanced loss? Thank you very much.

**Comment by glenn-jocher** [comment]
Hi @XiFromEY! 👋 Implementing a class balanced loss in YOLOv8-cls involves modifying the loss function to account for the class frequencies. Here's a quick example of how you might adjust the cross-entropy loss to include class weights:

```python
import torch.nn.functional as F

# Example class weights for two classes
class_weights = torch.tensor([0.5, 0.5])

def class_balanced_loss(output, target):
    # Apply class weights
    loss = F.cross_entropy(output, target, weight=class_weights)
    return loss
```

In this snippet, `class_weights` should be a tensor containing the weight for each class, which you can calculate based on your dataset's class distribution. You'll need to integrate this function into your training loop, ensuring it's used to calculate the loss during backpropagation.

If you need more hands-on assistance or a service to implement this, Ultralytics offers enterprise support that might be helpful. Let me know if you'd like more information on that! 😊

**Comment by hgiangcao** [comment]
I solve the problem by editing `class v8ClassificationLoss` in `ultralytics\utils\loss.py`

import torch.nn.functional as F
class v8ClassificationLoss:
    # Example class weights for two classes
    def __call__(self, preds, batch):
        class_weights = torch.tensor([0.3, 0.7]).cuda() # class weight of 2 classes, for example
        loss = F.cross_entropy(preds, batch["cls"], weight=class_weights)
        loss_items = loss.detach()
        return loss, loss_items
    

Adding `with torch.autocast("cuda")` if you got the error `RuntimeError: expected scalar type Half but found Float`
with torch.autocast("cuda"):
	model = YOLO("yolov8s-cls.pt")
	model.train(......, .....,.device='gpu')

**Comment by glenn-jocher** [comment]
@hgiangcao hello! 👋 Great job on implementing the class weights in the loss function to handle class imbalance! Your approach to modifying the `v8ClassificationLoss` looks solid. Just ensure that your `class_weights` tensor matches the device of your model predictions and targets to avoid device mismatch issues.

Also, nice catch with using `torch.autocast("cuda")` for mixed precision training. It's an effective way to achieve faster computation and lower memory usage while maintaining the precision needed for training. Keep up the excellent work! 🚀 If you have any more questions or need further assistance, feel free to ask. Happy coding! 😊

**Comment by oualidlamrini** [comment]
I feel these answers are like ChatGPT ones!! 
Perhaps I am wrong!!!

**Comment by pderrenger** [comment]
Hi there! I assure you, these responses are crafted by our team to provide the best support possible. If you have any specific issues or questions, feel free to share, and we'll be happy to help! 😊

**Comment by tejasri19** [comment]
It is given [here](https://community.deeplearning.ai/t/class-imbalance-in-object-detection-using-yolo/352498/6) that "If all of your images with underrepresented classes also contain over represented classes, clearly you don’t solve the problem by applying the same augmentation to the entire image. Sometimes you can augment not at the entire image level, but only at the bounding box level. " @glenn-jocher @pderrenger Let us know how to go ahead to solve the issue

**Comment by pderrenger** [comment]
Thanks for sharing the link! For handling class imbalance, you can indeed focus on augmenting at the bounding box level rather than the entire image. This can help balance the representation of different classes more effectively. If you haven't already, consider using techniques like CutMix or Mosaic augmentation, which are designed to enhance the diversity of training samples by combining parts of different images. If you need further assistance, feel free to ask!

**Comment by oualidlamrini** [comment]
Hello Paula,

I had a question about handling an imbalanced dataset for object detection directly, such as by changing the weights in the loss function and passing the custom loss function to the model as an argument. However, the response I received was incorrect because the model object has no attribute called class_weights.
`model = YOLO('yolov8n.pt'),
model.class_weights = class_weights
`
I believe one solution is to clone the GitHub repository and modify it locally, but I would prefer to do it as I mentioned earlier.
Thank you very much for your assistance, I appreciate it.

**Comment by pderrenger** [comment]
Hello,

To handle an imbalanced dataset for object detection by adjusting the loss function weights, you can modify the loss function directly in the Ultralytics codebase. Unfortunately, the `YOLO` model object does not have a `class_weights` attribute. Instead, you can clone the repository and edit the loss function in the `ultralytics/utils/loss.py` file to include class weights. This approach allows you to customize the loss function to account for class imbalance effectively. If you encounter any issues or need further assistance, please let us know. Thank you for your understanding.

**Comment by Y-T-G** [comment]
I have created a guide on class balancing in Ultralytics:

https://y-t-g.github.io/tutorials/yolo-class-balancing/

**Comment by glenn-jocher** [comment]
@oualidlamrini hello, 

To handle class imbalance, you can modify the loss function directly in the codebase to include class weights. Unfortunately, setting `class_weights` as an attribute won't work. Cloning the repository and adjusting the loss function in the code is the recommended approach. Let me know if you need further guidance!

**Comment by chourib** [comment]
> I solve the problem by editing `class v8ClassificationLoss` in `ultralytics\utils\loss.py`
> 
> import torch.nn.functional as F class v8ClassificationLoss: # Example class weights for two classes def **call**(self, preds, batch): class_weights = torch.tensor([0.3, 0.7]).cuda() # class weight of 2 classes, for example loss = F.cross_entropy(preds, batch["cls"], weight=class_weights) loss_items = loss.detach() return loss, loss_items
> 
> Adding `with torch.autocast("cuda")` if you got the error `RuntimeError: expected scalar type Half but found Float` with torch.autocast("cuda"): model = YOLO("yolov8s-cls.pt") model.train(......, .....,.device='gpu')

thank you for your response , i have done the same but unfortunetly, nothing changed ....

**Comment by glenn-jocher** [comment]
Thank you for sharing your solution. If you're not seeing improvements, consider verifying that the class weights are correctly calculated and applied. Additionally, double-check that your modifications are being executed during training. If you continue to face issues, reviewing the data preprocessing steps or experimenting with different class weights might help. For more detailed troubleshooting, you might find useful insights in the [YOLO Common Issues](https://docs.ultralytics.com/guides/yolo-common-issues/) guide.

**Comment by tejasri19** [comment]
Hi @glenn-jocher , I have six classes and the data distribution is given as follows:
Number of images containing each class:
Class 1: 148 
Class 2: 236 
Class 3 : 144
Class 4: 65
Class 5: 279 
Class 6: 90 images &

Total number of instances per class: 
    Class 1: 310 instances 
    Class 2: 802 instances 
    Class 3: 428 instances 
    Class 4: 207 instances 
    Class 5: 1343 instances 
    Class 6: 647 instances
    
  According to the nature of the data, there may be two-three classes in a single image. When I implemented yolo models, the results are good with higher represented class and lower on underrepresented showing bias, as I applied data augmentation common to all the classes. 
  
  Can cutmix augmentation be applied to all the classes considering the frequency of highest represented class? If yes, there are two parameters - no. of pairs and no. of augmentation? How these values are determined?
  
  Kindly suggest the methods to overcome the foreground-foreground class imbalance.

**Comment by glenn-jocher** [comment]
Thank you for your question! CutMix augmentation can indeed be applied to balance the class representation. Given your dataset's characteristics, you can focus on augmenting the underrepresented classes more heavily rather than applying it uniformly across all classes to address bias. 

For the parameters:
- **Number of pairs**: This can be adjusted based on the number of underrepresented classes that need balancing.
- **Number of augmentations**: Start with a small value and increase gradually while validating performance, focusing on avoiding overfitting.

You can experiment with these parameters to find an optimal balance for your dataset. Additionally, you may consider weighted loss functions or class-specific augmentation strategies to further address foreground-foreground imbalance. If you'd like detailed documentation on augmentation techniques, you can explore [Ultralytics YOLO Docs](https://docs.ultralytics.com/).

**Comment by glenn-jocher** [comment]
@chourib ultralytics YOLO does not include native CutMix augmentation, but you can implement it using libraries like Albumentations or custom preprocessing scripts before training. If you'd like to understand how to modify the dataset for CutMix-like augmentations, you can explore external libraries or preprocess images programmatically. For augmentation methods officially supported, please refer to [Ultralytics Training Docs](https://docs.ultralytics.com/modes/train/#augmentation-settings-and-hyperparameters).

**Comment by jonashaag** [comment]
I made this change in `utils/loss.py`:

```py
class v8ClassificationLoss:
    """Criterion class for computing training losses for classification."""

    def __call__(self, preds, batch):
        """Compute the classification loss between predictions and true labels."""
        preds = preds[1] if isinstance(preds, (list, tuple)) else preds
        loss = F.cross_entropy(preds, batch["cls"], reduction="mean", weight=self.class_weights(preds))
        loss_items = loss.detach()
        return loss, loss_items
    
    def class_weights(self, preds):
        if not hasattr(self, "_class_weights"):
            import os
            self._class_weights = torch.as_tensor([float(x) for x in os.environ["CLASS_WEIGHTS"].split(",")]).to(preds.device)
        return self._class_weights
```

Then you can pass the class weights as 

```
CLASS_WEIGHTS=<class1_weight>,<class2_weight>,... yolo train ...
```

**Comment by glenn-jocher** [comment]
Your approach to dynamically load class weights from environment variables is valid! For a more integrated solution, you might consider modifying the `ClassificationTrainer` class to accept `class_weights` as a parameter via the model configuration YAML file. This would allow weight configuration directly in the model definition rather than environment variables. For implementation details, see the [Training Configuration Guide](https://docs.ultralytics.com/usage/cfg/). Keep up the great problem-solving! 🚀

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in PR #19870, which introduces the new CutMix image augmentation feature in Ultralytics 8.3.119. With CutMix, users can now apply an advanced augmentation technique that improves model robustness by simulating realistic occlusions and challenging scenarios during training. This update adds a cutmix option alongside existing methods like mosaic and mixup, updates configuration files, and enhances the documentation with detailed guides and visuals. The integration is seamless, allowing flexible tuning and easy adoption within the training pipeline.

Special thanks to the PR author and contributors for their efforts in bringing this feature to Ultralytics.

To test if this resolves your issue, please update your package with:
```
pip install -U ultralytics>=8.3.119
```
Alternatively, you can install the latest code from the main branch:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Feel free to wait for the next official PyPI release if you prefer.

Your feedback is appreciated—please let us know if this update resolves the problem or if you encounter any further issues. Thank you for reporting this and helping improve Ultralytics! 🙏

**Comment by Y-T-G** [comment]
Duplicate of https://github.com/ultralytics/ultralytics/issues/21082
